### PR TITLE
Add kinto changes listener

### LINKED ISSUE #12: Support a mode where we don't push preview buckets
Right now we push a notification on e.g. every change to the monitor-changes bucket. That means we trigger when records are first copied into the main-preview bucket as well as after they are approved and copied into the main bucket. However, the version ID is the same in both cases, so megaphone doesn't treat the second push as an actual push. Instead, the client gets notified when the preview collection is updated. This isn't ideal.

Discussing this with Mat on IRC, here are some proposed approaches:

- Maybe the monitor/changes collection should get a new ETag when the records are copied to main. But this seems hard because the records should be the same records from the main-preview collection, and the ETag comes from the latest last_modified in the collection.
- Maybe we could remove the preview buckets from the monitor/changes resource. But this is inconvenient because the client relies on finding things in monitor/changes to load them, even for preview collections.
- Finally, maybe we can exclude the preview buckets from kinto-megaphone events. This means setting up a new type of listener that is somehow smart enough to decompose events on the kinto-changes collection, filter out records that are relevant to it, and perform kinto-megaphone updates only sometimes. This seems like the most plausible solution, although it requires a certain amount of work.

The filtering proposed here goes a bit beyond the notification filtering proposed in https://kinto.readthedocs.io/en/stable/tutorials/notifications-custom.html and https://github.com/Kinto/kinto/pull/1499, so maybe we should just build something into this package.

### PR DESCRIPTION
**Pull request by glasserc**
Fixes #12.

### CODE REVIEW

File: kinto_megaphone/listeners/kinto_changes.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+    def __call__(self, event):
+        if event.payload['resource_name'] != 'record':
+            logger.debug("Resource name did not match. Was: {}".format(
+                event.payload['resource_name']))
+            return
+
+        bucket_id = event.payload['bucket_id']
+        collection_id = event.payload['collection_id']
+        if bucket_id != 'monitor' or collection_id != 'changes':
```
[the remark this answers is not in the document]
Comment (glasserc) [reply kinto_megaphone/listeners/kinto_changes.py]: Good idea, thanks.

**Comment by glasserc** [comment]
Following up from a discussion I had with Mat. If you first come to this PR, you might expect a clean/elegant `exclude_changes` setting that filters events generically, something like https://github.com/Kinto/kinto/pull/1499. But the tricky part here is that we don't necessarily want to filter all events from the `monitor/changes` collection. Instead, we want to look inside those events, understand them as `kinto-changes` events, figure out what collections they represent changes to, and exclude some of those. This makes everything a bit more complicated/encapsulation-violating. I tried to document why in the docstrings/README but maybe I could have been more clear; in which case don't hesitate to make suggestions.

Another thing that came up when discussing this is whether we really need to have two listeners. Megaphone is pretty limited in scope to Firefox and Mozilla, and the only use case we have for it right now is the `monitor/changes` collection, so why bother keeping the `CollectionTimestampListener`? I agree with this point -- having just one listener makes things easier to configure and deploy -- so I'll merge them together (although probably keep the `KintoChangesListener` name).

File: kinto_megaphone/listeners/kinto_changes.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            if 'new' not in delta:
+                continue  # skip deletes
+            record = delta['new']
+            record_bucket = record['bucket']
+            record_collection = record['collection']
+            for (resource_name, matchdict) in self.resources:
+                if resource_name == 'bucket':
+                    resource_bucket = matchdict['id']
+                else:
+                    resource_bucket = matchdict.get('bucket_id')
```
[the remark this answers is not in the document]
Comment (glasserc) [reply kinto_megaphone/listeners/kinto_changes.py]: I guess so. In principle, we have other resources that don't belong in buckets, although `kinto-changes` doesn't watch them.

File: kinto_megaphone/listeners/kinto_changes.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+                if resource_name == 'collection':
+                    resource_collection = matchdict['id']
+                else:
+                    resource_collection = matchdict.get('collection_id')
+
+                if resource_bucket and resource_bucket != record_bucket:
+                    continue
```
[the remark this answers is not in the document]
Comment (glasserc) [reply kinto_megaphone/listeners/kinto_changes.py]: Why? In my mind, it was "Fetch bucket/collection from the resource, then check if the record matches".

File: tests/test_listeners_kinto_changes.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+    assert excinfo.value.args[0] == error_msg
+
+
+def test_kinto_changes_complains_about_missing_config_param(kinto_changes_settings):
+    del kinto_changes_settings['event_listeners.mp.match_kinto_changes']
+    config = Configurator(settings=kinto_changes_settings)
+    with pytest.raises(ConfigurationError) as excinfo:
+        load_from_config(config, 'event_listeners.mp.')
+    ERROR_MSG = "Resources to filter must be provided to kinto_changes" \
+        " using match_kinto_changes"
```
Comment (leplatrem) [inline tests/test_listeners_kinto_changes.py]: nit: we usually use parenthesis instead of backslash

File: README.rst [hunk 5]
```diff
@@ -46,9 +50,7 @@ Add it using configuration like::
 
   kinto.event_listeners = mp
   kinto.event_listeners.mp.use = kinto_megaphone.listeners
-
-Every listener also needs the following settings (with real values)::
-
   kinto.event_listeners.mp.api_key = foobar
   kinto.event_listeners.mp.url = http://megaphone.example.com/
   kinto.event_listeners.mp.broadcaster_id = remote-settings
+  kinto.event_listeners.mp.match_kinto_changes = /buckets/main /buckets/blocklists /buckets/gfx
```
Comment (leplatrem) [inline README.rst]: you could add an example where a collection is used.

(and a note that it doesn't work with groups ;))

**Comment by glasserc** [comment]
Another point that came up during discussion is that this plugin doesn't really care what's in the kinto-changes configuration. We don't match the `match_kinto_changes` parameter to the `kinto.changes.resources` in any way. There's no value in consulting the `kinto.changes.resources` configuration because we have to check against the complete resource we were told in any case.

**Pull request merged**